File: steamdeck_brain/vaelrix_forcefield/pixelbrain/pbrain_checksum.py
"""
PixelBrain .pbrain asset checksum — verifier.

The image-trace tracer stamps each PixelBrainAssetPacket with:

    "checksum": {
        "algorithm": "FNV-1a-32",
        "scope": "canonical JSON excluding this checksum object",
        "value": "<UPPER_HEX8>"
    }

The "canonical JSON" form is Python-compact: the top-level ``checksum`` key is
removed and the remaining packet is serialized with
``json.dumps(obj, separators=(",", ":"))`` (no indent, file key order). The
FNV-1a-32 (offset basis 2166136261, prime 16777619) of that UTF-8 string,
formatted as upper-case 8-hex, is the checksum value.

NOTE: this verification MUST run on a numeric-type-preserving reader. JSON
floats such as ``64.0`` serialize back as ``64.0`` in Python but as ``64`` in
JavaScript (and ``JSON.parse`` already discards the distinction), so a JS
verifier cannot reproduce the byte stream. Verify .pbrain checksums here.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def fnv1a32(text: str) -> str:
    """FNV-1a-32 of a string, as upper-case zero-padded 8-hex."""
    h = 2166136261
    for ch in text:
        h ^= ord(ch)
        h = (h * 16777619) & 0xFFFFFFFF
    return format(h, "08X")


def canonical_pbrain_json(packet: dict[str, Any]) -> str:
    """Canonical serialization the tracer checksums: packet minus checksum, compact."""
    body = {k: v for k, v in packet.items() if k != "checksum"}
    return json.dumps(body, separators=(",", ":"))


def compute_pbrain_checksum(packet: dict[str, Any]) -> str:
    """Recompute the FNV-1a-32 checksum value for a packet."""
    return fnv1a32(canonical_pbrain_json(packet))


def verify_pbrain_checksum(packet: dict[str, Any]) -> tuple[bool, str, str]:
    """
    Verify a packet's stamped checksum.

    Returns (ok, expected, recomputed).
    """
    expected = str((packet.get("checksum") or {}).get("value", ""))
    recomputed = compute_pbrain_checksum(packet)
    return recomputed == expected, expected, recomputed


def verify_pbrain_file(path: str | Path) -> tuple[bool, str, str]:
    """Load a .pbrain file (preserving numeric literal types) and verify it."""
    packet = json.loads(Path(path).read_text(encoding="utf-8"))
    return verify_pbrain_checksum(packet)


if __name__ == "__main__":
    import sys

    target = sys.argv[1] if len(sys.argv) > 1 else None
    if not target:
        print("usage: python -m vaelrix_forcefield.pixelbrain.pbrain_checksum <file.pbrain>")
        raise SystemExit(2)

    ok, expected, recomputed = verify_pbrain_file(target)
    status = "OK" if ok else "MISMATCH"
    print(f"{status} expected={expected} recomputed={recomputed} file={target}")
    raise SystemExit(0 if ok else 1)

File: codex/core/pixelbrain/pbrain-checksum.js
/**
 * .pbrain packet checksum — JavaScript verifier/stamper.
 *
 * Mirrors steamdeck_brain/vaelrix_forcefield/pixelbrain/pbrain_checksum.py:
 * the checksum value is FNV-1a-32 (upper 8-hex) of the packet minus its
 * top-level "checksum" key, serialized Python-compact (see canonical-json.js).
 *
 * Verification MUST go through `verifyPbrainText` / the *FromText variants,
 * which parse with numeric-literal preservation. Feeding a JSON.parse'd
 * object here would collapse floats (64.0 → 64) and produce a false mismatch
 * — the exact failure this module exists to prevent.
 */

import { readFileSync } from 'node:fs';
import { fnv1a32Hex } from './shared.js';
import { canonicalStringify, parseCanonicalJson } from './canonical-json.js';

export const PBRAIN_CHECKSUM_ALGORITHM = 'FNV-1a-32';
export const PBRAIN_CHECKSUM_SCOPE = 'canonical JSON excluding this checksum object';

function packetBody(packet) {
  if (packet instanceof Map) {
    const body = new Map();
    for (const [k, v] of packet) {
      if (k !== 'checksum') body.set(k, v);
    }
    return body;
  }
  if (packet && typeof packet === 'object' && !Array.isArray(packet)) {
    const { checksum, ...body } = packet;
    return body;
  }
  throw new TypeError('.pbrain packet must be a JSON object');
}

/**
 * Recompute the checksum value for a parsed packet (Map from
 * parseCanonicalJson, or a JS object using PyFloat markers for float fields).
 */
export function computePbrainChecksum(packet) {
  return fnv1a32Hex(canonicalStringify(packetBody(packet)));
}

/** Recompute the checksum value directly from raw .pbrain file text. */
export function computePbrainChecksumFromText(text) {
  return computePbrainChecksum(parseCanonicalJson(text));
}

/**
 * Verify raw .pbrain file text against its stamped checksum.
 * Returns { ok, expected, recomputed }.
 */
export function verifyPbrainText(text) {
  const packet = parseCanonicalJson(text);
  if (!(packet instanceof Map)) {
    throw new TypeError('.pbrain packet must be a JSON object');
  }
  const stamp = packet.get('checksum');
  const expected = String((stamp instanceof Map ? stamp.get('value') : stamp?.value) ?? '');
  const recomputed = computePbrainChecksum(packet);
  return { ok: recomputed === expected, expected, recomputed };
}

/** Build the checksum stamp object for a packet body. */
export function stampPbrainChecksum(packet) {
  return {
    algorithm: PBRAIN_CHECKSUM_ALGORITHM,
    scope: PBRAIN_CHECKSUM_SCOPE,
    value: computePbrainChecksum(packet),
  };
}

/** Thrown when a .pbrain file cannot be shown to be what it claims to be. */
export class PbrainIntegrityError extends Error {
  constructor({ path, reason, expected, recomputed, cause }) {
    const detail = {
      unreadable: `cannot be read: ${cause?.message ?? 'unknown error'}`,
      absent: 'carries no checksum stamp, so its contents cannot be verified',
      mismatch: `does not match its stamp (declares ${expected}, contents hash to ${recomputed})`,
    }[reason];
    super(`.pbrain integrity failure — ${path} ${detail}`);
    this.name = 'PbrainIntegrityError';
    this.path = path;
    this.reason = reason;
    this.expected = expected;
    this.recomputed = recomputed;
    if (cause) this.cause = cause;
  }
}

/**
 * Read a .pbrain file, PROVE it is what it claims to be, then parse it.
 *
 * Every load site used to be `JSON.parse(readFileSync(path))`, which cannot
 * verify anything: JSON.parse collapses 64.0 to 64, so the digest can only
 * ever be computed from the raw text. Verification therefore happens here,
 * on the bytes, before the packet is handed to a consumer.
 *
 * Strict by construction. A missing stamp throws exactly like a wrong one —
 * otherwise deleting the checksum key would be an easier bypass than forging
 * it, and the check would be one a corrupt packet could always pass.
 *
 * @returns the parsed packet (plain JSON objects, as callers already expect)
 * @throws {PbrainIntegrityError}
 */
export function loadPbrainFile(path) {
  let text;
  try {
    text = readFileSync(path, 'utf8');
  } catch (cause) {
    throw new PbrainIntegrityError({ path, reason: 'unreadable', cause });
  }

  const { ok, expected, recomputed } = verifyPbrainText(text);
  if (!ok) {
    throw new PbrainIntegrityError({
      path,
      reason: expected === '' ? 'absent' : 'mismatch',
      expected,
      recomputed,
    });
  }

  return JSON.parse(text);
}
